GPII-2486: Improve code coverage report...

the-t-in-rtf opened the pull request:
See [GPII-2486](https://issues.gpii.net/browse/GPII-2486) for details.

the-t-in-rtf commented:
See http://the-t-in-rtf.github.io/coverage-reports/universal/20170804/ for a sample updated coverage report.

the-t-in-rtf commented:
The test job fails with errors about not being able to find `fluid.model.transform`.

```
not ok 112 Firefox 54.0 - Global error: TypeError: fluid.model.transform is undefined at http://localhost:7357/201/node_modules/infusion/src/framework/core/js/DataBinding.js, line 688
    ---
        Log: |
            { type: 'info',
              text: '\'MOCK GET: /available-authorized-preferences/1\' { url: \'/available-authorized-preferences/1\',\n  type: \'GET\',\n  isLocal: false,\n  global: true,\n  processData: true,\n  async: true,\n  contentType: \'application/x-www-form-urlencoded; charset=UTF-8\',\n  accepts: \n   { \'*\': \'*/*\',\n     text: \'text/plain\',\n     html: \'text/html\',\n     xml: \'application/xml, text/xml\',\n     json: \'application/json, text/javascript\',\n     script: \'text/javascript, application/javascript, application/ecmascript, application/x-ecmascript\' },\n  contents: { xml: {}, html: {}, json: {}, script: {} },\n  responseFields: \n   { xml: \'responseXML\',\n     text: \'responseText\',\n     json: \'responseJSON\' },\n  converters: { \'text html\': true },\n  flatOptions: { url: true, context: true },\n  jsonp: \'callback\',\n  dataType: \'json\',\n  method: \'GET\' }\n' }
            { type: 'info',
              text: '\'MOCK GET: /authorizations/10/preferences\' { url: \'/authorizations/10/preferences\',\n  type: \'get\',\n  isLocal: false,\n  global: true,\n  processData: true,\n  async: true,\n  contentType: \'application/x-www-form-urlencoded; charset=UTF-8\',\n  accepts: \n   { \'*\': \'*/*\',\n     text: \'text/plain\',\n     html: \'text/html\',\n     xml: \'application/xml, text/xml\',\n     json: \'application/json, text/javascript\',\n     script: \'text/javascript, application/javascript, application/ecmascript, application/x-ecmascript\' },\n  contents: { xml: {}, html: {}, json: {}, script: {} },\n  responseFields: \n   { xml: \'responseXML\',\n     text: \'responseText\',\n     json: \'responseJSON\' },\n  converters: { \'text html\': true },\n  flatOptions: { url: true, context: true },\n  jsonp: \'callback\',\n  dataType: \'json\',\n  method: \'get\' }\n' }
            { type: 'error',
              text: 'TypeError: fluid.model.transform is undefined at http://localhost:7357/201/node_modules/infusion/src/framework/core/js/DataBinding.js, line 688\n' }
    ...
not ok 113 Firefox 54.0 - Edit privacy settings: Verify the rendering and closing of the edit dialog
    ---
        stack: >
            http://localhost:7357/201/node_modules/infusion/src/framework/core/js/DataBinding.js:688
        message: >
            TypeError: fluid.model.transform is undefined
        Log: |
    ...

...

not ok 122 Firefox 54.0 - GPII OAuth2 Component Tests: ../../../gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/webTests/all-tests.html
    ---
        stack: >
            runLoggingCallbacks@http://localhost:7357/201/node_modules/infusion/tests/lib/qunit/js/qunit.js:1609:4
            push@http://localhost:7357/201/node_modules/infusion/tests/lib/qunit/js/qunit.js:958:3
            onIframeLoad/<@http://localhost:7357/201/node_modules/infusion/tests/lib/qunit/addons/composite/qunit-composite.js:87:8
            runLoggingCallbacks@http://localhost:7357/201/node_modules/infusion/tests/lib/qunit/js/qunit.js:1609:4
            pushFailure@http://localhost:7357/201/node_modules/infusion/tests/lib/qunit/js/qunit.js:996:3
            window.onerror@http://localhost:7357/201/node_modules/infusion/tests/lib/qunit/js/qunit.js:1242:4
            interceptWindowOnError/window.onerror@http://localhost:7357/testem.js:1054:7
            
        message: >
            Component Tests: ./components/privacySettings/html/PrivacySettings-test.html: Edit privacy settings: Verify the rendering and closing of the edit dialog: TypeError: fluid.model.transform is undefined
        Log: |
    ...
```

Seems like there are still some stability issues, but I can't reproduce them locally.

amb26 commented:
No stability issues, just a missing include, as before

https://github.com/GPII/universal/blob/master/gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/webTests/components/privacySettings/html/PrivacySettings-test.html#L21

the-t-in-rtf commented:
@amb26, not sure what you mean there.  That include doesn't seem to be missing for me, either when hosting the tests in python and opening them in a browser, or when running through Testem.  Can I get more details about when you're encountering this?

amb26 commented:
Sorry, I was just highlighting the place where the includes are missing. Here is the standard sequence at that point from one of our component tests:
https://github.com/fluid-project/infusion/blob/master/tests/component-tests/pager/html/Pager-test.html#L14

the-t-in-rtf commented:
So, I'm looking for some objective feedback about the report itself, whether it is "good enough" for now.  It turns out I wasn't actually managing to combine the browser tests with anything else, mainly because istanbul and nyc can't quite interoperate in the way we need them to.

First of all, as a reminder we have to use nyc for the browser tests because istanbul cannot successfully launch testem.  This appears to set us up for compatibility issues between the two.

Istanbul can instrument whole directories with a `--complete-copy` option, which is well suited to our mix of code and configuration.  Unfortunately, although nyc's comparable instrument functionality promises to provide the same support, I haven't managed to convince it to instrument the `gpii` subdirectory yet, or come up with a workaround.

So, if I try to use nyc to report on coverage data generated in a browser using Istanbul-instrumented code, I get errors because the results lack the `decl` variable added by nyc:

```
[vagrant@universal universal]$ node node_modules/nyc/bin/nyc.js report --temp-directory=./coverage -r lcov
/home/vagrant/sync/universal/node_modules/nyc/node_modules/istanbul-reports/lib/lcovonly/index.js:32
        writer.println('FN:' + [meta.decl.start.line, meta.name].join(','));
```
The solutions I've found in various forums suggest that you need to use nyc to instrument and report on the resulting code coverage.   can't get nyc to correctly instrument our nested node_modules, so we're kind of stuck at the moment.

So, why is the report working at all?  I didn't realize that nyc was using its `.nyc_output` directory instead of the `coverage` directory used by Testem.  As a result, I was basically only reporting on the node fixtures used in the browser tests and the node tests.

Our choices at the moment appear to be:

1. Use nyc to report on the code coverage of the node fixtures used in the browser tests, and the node unit tests.
2. Use Istanbul to report on the browser coverage and node unit tests, with improved filtering.

Of the two, the first seems to work better with our particular test suite.  Anyway, let's leave this open until I return, with luck there will be a release of istanbul and/or nyc that will offer us new options.  Or I might have new ideas myself.

the-t-in-rtf commented:
@cindyli left [a comment on the associated JIRA ticket](https://issues.gpii.net/browse/GPII-2486?focusedCommentId=27805) indicating that all three types of coverage are needed.  I will need to revisit this and see if there are any new options a couple of weeks later.

the-t-in-rtf commented:
@amb26, the dependencies are in a different order, and it's difficult to see where the "standard sequence" stops (I'm assuming everything between the line you linked to and the "ToolTip" include).  comparing that to what I have, I see that `jquery.ui` is not included and that the Model Transformation API is not included.  Everything else is included, but in a different order.  My first question is, which (if any) of these are you suggesting is "missing"?

My second question is, why is the "missing" include needed at all?  I see no errors loading this content, and the tests pass.  Does the "missing" include somehow change the behavior of the code being tested?

amb26 commented:
@the-t-in-rtf - the model transformation files are a required dependency of the framework - at least when loading a component which defines a model relay rule operating a transform. The mechanism for the failure is the same as that we saw with https://issues.gpii.net/browse/GPII-2456 - if people write tests that don't actually exercise *all* the functionality of the component, and then they miss out some necessary includes that are only loaded late in the component lifecycle, they can end up with race conditions like these. In detail, the mechanism is like this - 

If the tests which are written against the component conclude quickly (or synchronously), the IoC Testing framework will normally manage to tear down the entire component fixture before it ends up trying to operate functionality which depends on the missing includes. If, however, the system is heavily loaded, the teardown may happen late enough that the component can trigger a failure by trying to finish its startup (in this case, to process a model relay rule). People testing the system locally on powerful or lightly loaded machines will seem to see that the tests pass fine.

the-t-in-rtf commented:
Thanks for the explanation. I will add the model relay includes.

On Mon, Aug 28, 2017 at 4:17 PM, Antranig Basman <notifications@github.com>
wrote:

> @the-t-in-rtf <https://github.com/the-t-in-rtf> - the model
> transformation files are a required dependency of the framework - at least
> when loading a component which defines a model relay rule operating a
> transform. The mechanism for the failure is the same as that we saw with
> https://issues.gpii.net/browse/GPII-2456 - if people write tests that
> don't actually exercise *all* the functionality of the component, and
> then they miss out some necessary includes that are only loaded late in the
> component lifecycle, they can end up with race conditions like these. In
> detail, the mechanism is like this -
>
> If the tests which are written against the component conclude quickly (or
> synchronously), the IoC Testing framework will normally manage to tear down
> the entire component fixture before it ends up trying to operate
> functionality which depends on the missing includes. If, however, the
> system is heavily loaded, the teardown may happen late enough that the
> component can trigger a failure by trying to finish its startup (in this
> case, to process a model relay rule). People testing the system locally on
> powerful or lightly loaded machines will seem to see that the tests pass
> fine.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GPII/universal/pull/537#issuecomment-325366200>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFf4ZlJ7GD-Kg40BiRnq6UtaBmyaEDiQks5scswRgaJpZM4Otcil>
> .
>

the-t-in-rtf commented:
The model relay includes are now there (and in a few other spots that lacked them), see the commit.  I still need to come up with a working approach to combining the code coverage before this is ready for further review.

amb26 commented:
Thanks for the updates to includes, @the-t-in-rtf - there were typos in a couple of them (trailing full stop in filename), I've updated universal with corrected versions so please merge this branch up with master

the-t-in-rtf commented:
Thanks, @amb26.  Just to confirm, this is not ready for further review, I am still trying to find a way to collect coverage regarding all three parts (browser tests, server fixtures associated with browser tests, node tests).

the-t-in-rtf commented:
Hi, all.  I just finished discussing this with Antranig and wanted to summarise for the group.  In short, although the browser tests are greatly improved, I still have not found a way to create a combined code coverage report for browser tests, server-side fixtures hit from browser-tests, and node-only tests.  I can collect coverage data for all three, but not in a way where either of the available reporting tools (istanbul and its successor nyc) can work with them.

The sticking point is our use of the nested node_modules structure within this package, which it seems impossible to configure nyc to work with.  The same structure causes problems in other areas as well, and we know that long term we'll have to move to a monorepo pattern to address some of these.  Once this occurs, we should be able to use nyc with no problems, as we do in various other packages.

The question is, what to do in the short term.  The best I have managed thus far is to generate partially broken combined reports with accurate coverage data by using istanbul to collect the coverage data and using nyc to create the report.  Here is a sample:

http://the-t-in-rtf.github.io/coverage-reports/universal/20180115/index.html

Take a look at this entry in particular:

http://the-t-in-rtf.github.io/coverage-reports/universal/20180115/contextManager/src/ContextManager.js.html

As far as I can see, this is the result of using the instanbul-collected coverage data with nyc.  Once the coverage data for the node tests and Testem fixtures are collected using nyc, this should no longer be a problem.

Istanbul fails to create the report because the browser tests use code instrumented with a newer version of the instrumentation library.  I will set aside a small amount of time to see if I can make this work tomorrow.

Failing that, I would propose documenting the slight brokenness of the report as a known issue and revisiting once we move to a monorepo.

the-t-in-rtf commented:
So, the good news is that I finally got the combined coverage report working, including all content we want to see in the report, and excluding everything we don't.

However, before this can proceed, we need a fix for the failing transform that results in output like:

```
10:57:40.357:  jq: FAIL: Module "gpii.matchMakerFramework tests" Test name "inferCommonTerms Tests" - Message: Testing with application specific only settings
10:57:40.357:  jq: Expected: {
    "gpii-default": {
        "com.microsoft.windows.mouseTrailing": {
            "http://registry.gpii.net/common/mouseTrailing": 10
        },

...

10:57:40.357:  jq: Actual: {
    "gpii-default": {
        "com.microsoft.windows.mouseTrailing": {
            
        },
```

I know @amb26 and @kaspermarkus have been talking about this, happy to fix this with a little guidance if no one else is already working on it.

the-t-in-rtf commented:
Here's a sample of the combined report:

https://the-t-in-rtf.github.io/coverage-reports/linux/20180215/

the-t-in-rtf commented:
@amb26, this has been merged with master and the unique instances of `%universal` brought in by this pull have been updated to `%gpii-universal`.

amb26 commented:
ok to test

the-t-in-rtf commented:
Looks like it simply timed out on the browser tests?  Let's assume a one-off problem and try again.

ok to test.

amb26 commented:
It is not a timeout - the build has failed 3 times deterministically with the following failure:


    not ok 1 Firefox 58.0 - Global error: TypeError: iframeWin.QUnit is undefined at

the-t-in-rtf commented:
Thanks, I see that.  Updating my FF to 58 locally and seeing if I can reproduce.

the-t-in-rtf commented:
Works fine with FF 58 on OS X.  Updating a local copy of the same Vagrant VM and trying there.

the-t-in-rtf commented:
It works fine with `vagrant destroy -f && vagrant up && npm run test:vagrantBrowser`.  Trying with a clean VM and the same `vagrant ssh -c 'cd /home/vagrant/sync/universal; DISPLAY=:0; npm run test:browser'` command used in the build.

the-t-in-rtf commented:
OK, `vagrant destroy -f && vagrant up && vagrant ssh -c 'cd /home/vagrant/sync/universal; DISPLAY=:0; npm run test:browser'` also works fine.

the-t-in-rtf commented:
OK, as a theory I tried blowing away my host `node_modules` as well before recreating the VM.  Same result.  I cannot reproduce the CI failures.

amatas commented:
Hi @the-t-in-rtf and @amb26, I got the same error shown at the CI a my local PC. This is the error that I can see in the web browser of the VM:

![image](https://user-images.githubusercontent.com/158641/36984547-6ab062e2-2095-11e8-8544-d52b90cec9c4.png)

the-t-in-rtf commented:
ok to test.

the-t-in-rtf commented:
OK, the failure was a problem with the production tests (which I haven't really touched).

amb26 commented:
Yes, this problem is well known to @amatas , @gtirloni  and @cindyli  and will be resolved by replacing curl with wget

the-t-in-rtf commented:
So, is there work already being done on this that I can pull in, or is it enough to demonstrate that the tests I'm changing work?

gtirloni commented:
I'd try to merge master. It has a [change](https://github.com/GPII/universal/commit/0ada0d781c172935f3ee06b3c7fe69fc8fb5e79e) that makes this error less likely.

the-t-in-rtf commented:
Thanks, @gtirloni, merging and confirming working now.

amatas commented:
I've added this PR [GPII-2885](https://github.com/GPII/universal/pull/593) in order to use wget instead of curl.

amb26 commented on tests/web/html/all-tests.html:
```diff
@@ -11,25 +11,25 @@
-        "../../../gpii/node_modules/gpii-oauth2/gpii-oauth2-utilities/test/html/OAuth2UtilitiesTests.html",
-        "../../../gpii/node_modules/processReporter/test/web/html/ProcessesBridgeTest.html",
-        "../../../gpii/node_modules/processReporter/test/web/html/ProcessReporterTest.html"
+        "/gpii/node_modules/matchMakerFramework/test/html/MatchMakerUtilitiesTest.html",
```
This change means that I can no longer run the web-based tests from my local HTTP server - is there any way we can do without it?

the-t-in-rtf replied:
I haven't found a way yet, no.  The parent directory calls end up bypassing the routing that loads the instrumented code.  You can run the tests from your local HTTP server as long as you host universal as the root.

the-t-in-rtf replied:
I can make a separate fixture for testem if you like.

amb26 replied:
I don't suppose there's a way to detect whether we are running in testem or not, and issue the appropriate form of path?

the-t-in-rtf replied:
We have also discussed detecting the presence of the global Testem object and using that to choose which paths to use.

amb26 commented on tests/testem.js:
```diff
@@ -24,23 +24,49 @@ gpii.tests.universal.testem.getTestemOptions = function (that) {
-        }
+        tap_quiet_logs: true,
+        disable_watching: true,
+        skip: "PhantomJS,Opera,Safari"
```
I'd like to see IE excluded from this too since our only remaining web tests within universal don't test any user-delivered UIs, and the testem bugs wrt. IE are annoying

the-t-in-rtf replied:
Done.

amb26 commented:
This branch runs well on my machine. I notice that the reports for reports/universal/gpii/node_modules/testing/src/Testing.js.html show it to be largely uncovered, although I know it is used extensively by the acceptance tests

the-t-in-rtf commented:
We are still discussing the ../../.../gpii paths versus /gpii at the moment.  If at all possible, we want to keep the ability to host the repo with a web server and open the tests with a browser there.

the-t-in-rtf commented:
Based on discussions with @amb26 I came up with a simple way of changing the path based on the presence of the global Testem object.  I have confirmed working locally both within Testem, and when hosting the tests in Python and opening them directly in a browser.

the-t-in-rtf commented:
So, this is largely in good shape, although I still need to address [Antranig's comment regarding missing test coverage](https://github.com/GPII/universal/pull/537#issuecomment-370779039).  I  have that on my calendar for tomorrow.

the-t-in-rtf commented:
@amb26, I started searching for code that hits Testing.js.  If you have a minute, I could use help finding an example of an acceptance test that should exercise this.

amb26 commented:
Hi @the-t-in-rtf - you can straightforwardly run a standard acceptance test, e.g. via

    node tests/platform/windows/windows-builtIn-testSpec.js builtIn

Inserting exploding material e.g. within gpii.test.operateSettings shows that its body is indeed executed many times during the test run - including (hopefully) all the rest of the file contents

the-t-in-rtf commented:
Thanks for clarifying, @amb26. Only the `cloud` platform tests are run at the moment, which would seem to explain the gap.  I'm thinking through how to close this gap without inserting platform-specific tests into the general `npm test` script or otherwise reducing the "universality" of the tests.

As our Vagrant VM could run at least the linux platform acceptance tests, one option would be to simply run the linux platform tests as part of the Vagrant tests.  For example, I could add a `test:vagrantAcceptance` script and add that to the `test:vagrant` rollup.

We could also make a platform-detecting rollup for acceptance tests that always runs the cloud tests and that runs the additional tests for whichever platform is detected.  Of course, this would mainly be a convenience for individual developers, as we aren't really ready to talk about doing anything like collecting coverage in multiple platforms and then somehow preparing a combined report.

Anyway, please comment on which (if any) appeals, I can certainly update this today.

amb26 commented:
Thanks, this is indeed an interesting discrepancy. It *used* to be the case that at the very least, we ran all the "builtIn" tests on each platform as part of a test run (when we ran these manually) and these have got lost as part of the move to vagrant. We should just restore these to all-tests.
The tests should not be platform-specific - that is, they should not make any use of platform detection in deciding what tests to run. There has been elaborate and substantial work on mocking to ensure that the same tests run as "Integration" tests using mocks when they detect they are in a test environment - and so the run 
    node tests/platform/windows/windows-builtIn-testSpec.js builtIn
for example should run on each platform.

the-t-in-rtf commented:
Thanks for clarifying, @amb26.  I will restore those shortly.

the-t-in-rtf commented:
OK, based on feedback from @amb26, I added the platform-specific acceptance tests to the all-tests rollup.  I also fixed an nyc configuration oversight that prevented full coverage of the Testing.js file from being collected.  This should be ready for further review.  A copy of the latest coverage report can be found here:

http://the-t-in-rtf.github.io/coverage-reports/universal/20180309/

amb26 commented:
ok to test
